agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+                    Ok(_) => trace!("stop_all_discovery - discovery client for protocol {} at endpoint {} told to stop", self.config.spec.protocol.name, endpoint),
+                    Err(e) => error!("stop_all_discovery - discovery client for protocol {} at endpoint {} could not receive stop message with error {:?}", self.config.spec.protocol.name, endpoint, e)
+                }
+            }
+        }
+    }
+
+    /// Calls discover on the Discovery Handler at the given endpoint and returns the connection stream.
+    pub async fn get_stream(
+        &self,
+        endpoint: &str,
+        discovery_handler_path: &str,
+    ) -> Option<StreamType> {
+        let discover_request = tonic::Request::new(DiscoverRequest {
+            discovery_details: self.config.spec.protocol.discovery_details.clone(),
+        });
+        trace!("get_stream - endpoint is {}", endpoint);
+        if endpoint == EMBEDDED_DISCOVERY_HANDLER_ENDPOINT
+            && get_discovery_handler(&self.config.spec.protocol.discovery_details).is_ok()
+        {
+            trace!(
+                "get_stream - internal discovery handler for protocol {}",
+                self.config.spec.protocol.name
+            );
+            let discovery_handler =
+                get_discovery_handler(&self.config.spec.protocol.discovery_details).unwrap();
+            Some(StreamType::Internal(
+                discovery_handler
+                    .discover(discover_request)
+                    .await
+                    .unwrap()
+                    .into_inner(),
+            ))
+        // Check if is a UDS connection
+        } else if endpoint.starts_with(discovery_handler_path) {
+            let path = endpoint.to_string();
+            match Endpoint::try_from("lttp://[::]:50051")
+                .unwrap()
+                .connect_with_connector(tower::service_fn(move |_: Uri| {
+                    let endpoint = path.clone();
+                    tokio::net::UnixStream::connect(endpoint)
+                }))
+                .await
+            {
+                Ok(channel) => {
+                    trace!(
+                        "get_stream - external discovery handler for protocol {}",
+                        self.config.spec.protocol.name
+                    );
+                    let mut discovery_client = DiscoveryClient::new(channel);
+                    Some(StreamType::External(
+                        discovery_client
+                            .discover(discover_request)
+                            .await
+                            .unwrap()
+                            .into_inner(),
+                    ))
+                }
+                Err(e) => {
+                    error!("get_stream - failed to connect to client with error {}", e);
+                    None
+                }
+            }
+        } else {
+            match DiscoveryClient::connect(endpoint.to_string()).await {
```

Comment:
While discovery handlers always have to register with the Agents unix socket, the Agent supports connecting with the Discovery Handler over uds or IP. The use case for IP is narrow but could be helpful for the scenario where someone may want to run a Discovery Handler outside the cluster. They would need to deploy some sort of job to the cluster that would register the external Discovery Handler, specifying the IP endpoint during registration. Then the Agent would call the external one.